kernel/src/capability_ptr.rs

```diff
@@ -7,26 +7,23 @@
 use core::fmt::{Formatter, LowerHex, UpperHex};
 use core::ops::AddAssign;
 
-/// A pointer with target specific metadata concerning validity or access rights.
+/// A pointer to userspace memory.
```

Comment:
Hmm, I don't think it is necessary to be userspace memory. This may also find internal use.

Reply:
Can you give an example? Would it feel better if this is memory userspace can refer to (e.g.., it may not be in a processes memory region, and might in principle be shared from kernel memory or another process).

I think it's really important to nail down what exactly this abstraction represents. We can use capability hardware to implement different abstractions if we want.

Reply:
I think a capability could legally refer to any memory, user or otherwise. Once capabilities reach drivers, they may very well point to either kernel memory or user memory. I may wish to have the same driver write into userspace or an internal buffer depending on some mux layer below it.

Specifically in the case of user memory (because of how we upgrade the syscalls interface) it might be true to say that "any capability referring to user memory has been checked for validity", but that's about it.

Reply:
I think that drivers writing into userspace memory is a can of worms we should not open in the scope of this PR. :smile: I know that CHERI provides us potentially more assurances and flexibility than we used to have on our conventional targets, but we deliberately not allow such behavior in upstream Tock.

> Specifically in the case of user memory (because of how we upgrade the syscalls interface) it might be true to say that "any capability referring to user memory has been checked for validity", but that's about it.

With the changes discussed this Friday, that is no longer true, right? Regardless, that's a strong assertion we'll want to be careful about making.

Reply:
I think it's important to disentangle intended use-case from guarantees here.

According to my understanding a `CapabilityPtr`---right now---is
- able to hold arbitrary pointers and associated metadata,
- intended to work for providing pointers with associated metadata to userspace, and encompass such pointers passed from userspace,
- where the `CapabilityPtr` type does not in itself give any guarantees over the validity of the pointer its metadata whatsoever,
- but constructing one with arbitrary metadata is potentially unsafe as it can convey authority to access a region of memory.

I think the documentation needs to reflect these guarantees, use-cases, and safety concerns.